Changelog for the 2.0.0 release (#607)
rmccue/Requests: CHANGELOG.md

  For the lifetime of Requests 2.x, the deprecation notices can be disabled by defining a global `REQUESTS_SILENCE_PSR0_DEPRECATIONS` constant and
setting the value of this constant to `true`.

  A complete "translation table" between the Requests 1.x and 2.x class names is available in the [upgrade guide](https://requests.ryanmccue.info/docs/upgrading.html).

  Users of the Requests native custom autoloader will need to adjust their code to initialize the autoloader:
  ```php
  // OLD: Using the custom autoloader in Requests 1.x.
  require_once 'path/to/Requests/library/Requests.php';
  Requests::register_autoloader();

  // NEW: Using the custom autoloader in Requests 2.x.
  require_once 'path/to/Requests/src/Autoload.php';
  WpOrg\Requests\Autoload::register();
  ```

  (props [@jrfnl][gh-jrfnl], [@schlessera][gh-schlessera], [#503][gh-503], [#519][gh-519], [#586][gh-586], [#587][gh-587], [#594][gh-594])

- **A large number of classes have been marked as `final`**

  Marking a class as `final` prohibits extending it.

>   These changes were made after researching which classes were being extended in userland code and due diligence has been applied before making these changes. If this change is causing problem we didn't anticipate, please [open an issue to report it](https://github.com/WordPress/Requests/issues/new/choose).

Author: ```suggestion
  These changes were made after researching which classes were being extended in userland code and due diligence has been applied before making these changes. If this change is causing a problem we didn't anticipate, please [open an issue to report it](https://github.com/WordPress/Requests/issues/new/choose).
```

  If you still find the input validation to be too strict and you have a good use-case of why it should be loosened for a particular entry point, please [open an issue to discuss this](https://github.com/WordPress/Requests/issues/new/choose).

  The code within Requests itself has also received various improvements to be more type safe.

  (props [@jrfnl][gh-jrfnl], [@schlessera][gh-schlessera], [#499][gh-499], [#542][gh-542], [#547][gh-547], [#558][gh-558], [#572][gh-572], [#573][gh-573], [#574][gh-574], [#591][gh-591], [#592][gh-592], [#593][gh-593], [#601][gh-601], [#602][gh-602], [#603][gh-603], [#604][gh-604], [#605][gh-605], [#609][gh-609], [#610][gh-610], [#611][gh-611], [#613][gh-613], [#614][gh-614], [#615][gh-615], [#620][gh-620], [#621][gh-621], [#629][gh-629])

- **Update bundled certificates as of 2021-10-26**

  The bundled certificates were updated.

  Previously the bundled certificates in Requests would include a small subset of expired certificates for legacy reasons.
  This is no longer the case as of Requests 2.0.0.

>   :warning: **Note**: the included certificates bundle is only intended as a fall-back.

Author: ```suggestion
  :warning: **Note**: the included certificates bundle is only intended as a fallback.
```

  Setting the `$options['verify']` key to `true` when initiating a request enables certificate verification using the certificate authority list provided by the server environment, which is recommended.

  The [documentation regarding Secure Requests with SSL](https://requests.ryanmccue.info/docs/usage-advanced.html#secure-requests-with-ssl) has also been updated to reflect this and it is recommended to have a read through.

  The included certificates _file_ has now also been moved to a dedicated `/certificates` directory off the project root.

  (props [@jrfnl][gh-jrfnl], [@schlessera][gh-schlessera], [@wojsmol][gh-wojsmol], [@ZsgsDesign][gh-ZsgsDesign], [#535][gh-535], [#571][gh-571], [#577][gh-577], [#622][gh-622], [#632][gh-632])

- **New functionality**

  The following new functionality has been added:
  - A `public static` `WpOrg\Requests\Requests::has_capabilities($capabilities = array())` method is now available to check whether there is a transport available which supports the requested capabilities.

>   - A `public` `WpOrg\Requests\Response::decode_body($associative = true, $depth = 512, $options = 0)` method is now available to handle json decoding a response body.

Author: JSON is an acronym (JavaScript Object Notation).
```suggestion
  - A `public` `WpOrg\Requests\Response::decode_body($associative = true, $depth = 512, $options = 0)` method is now available to handle JSON-decoding a response body.
```

    The method parameters correspond to the parameters of the PHP native [`json_decode()`](https://php.net/json-decode) function.

> 	The method will throw an `WpOrg\Requests\Exception` when the response body is not valid json.

Author: ```suggestion
	The method will throw an `WpOrg\Requests\Exception` when the response body is not valid JSON.
```

>   - A `WpOrg\Requests\Capability` interface. This class encapsulates the known capabilities. Transports can be tested whether or not they support these capabilities.

Author: ```suggestion
  - A `WpOrg\Requests\Capability` interface. This interface provides constants for the known capabilities. Transports can be tested whether or not they support these capabilities.
```

    Currently, the only capability supported is `Capability::SSL`.
  - A `WpOrg\Requests\Port` class. This class encapsulates typical port numbers as constants and offers a `static` `Port::get($type)` method to retrieve a port number based on a request type.
    Using this class when referring to port numbers is recommended.
  - An `WpOrg\Requests\Exceptions\InvalidArgument` class. This class is intended for internal use only.

>   - An `WpOrg\Requests\Utility\InputValidator` class with helper method for input validation. This class is intended for internal use only.

Author: ```suggestion
  - An `WpOrg\Requests\Utility\InputValidator` class with helper methods for input validation. This class is intended for internal use only.
```

  (props [@ccrims0n][gh-ccrims0n], [@dd32][gh-dd32], [@jrfnl][gh-jrfnl], [@schlessera][gh-schlessera], [#167][gh-167], [#214][gh-214], [#250][gh-250], [#251][gh-251], [#492][gh-492], [#499][gh-499], [#538][gh-538], [#542][gh-542], [#547][gh-547], [#559][gh-559])

- **Changed functionality**

  - The `WpOrg\Requests\Requests::decompress()` method has been fixed to recognize more compression levels and handle these correctly.
  - The method signature of the `WpOrg\Requests\Transport::test()` interface method has been adjusted to enforce support for an optional `$capabilities` parameter.

>     The Request native `WpOrg\Requests\Transport\Curl::test()` and `WpOrg\Requests\Transport\Fsockopen::test()` methods, both already supported this parameter.

Author: ```suggestion
    The Request native `WpOrg\Requests\Transport\Curl::test()` and `WpOrg\Requests\Transport\Fsockopen::test()` methods both already supported this parameter.
```

  - The `WpOrg\Requests\Transport\Curl::request()` and the `WpOrg\Requests\Transport\Fsockopen::request()` methods will now throw an `WpOrg\Requests\Exception` when the `$options['filename']` contains an invalid path.
  - The `WpOrg\Requests\Transport\Curl::request()` method will no longer set the `CURLOPT_REFERER` option.
  - The default value of the `$key` parameter in the `WpOrg\Requests\Cookie\Jar::normalize_cookie()` method has been changed from `null` to an empty string.

  (props [@datagutten][gh-datagutten], [@dustinrue][gh-dustinrue], [@jrfnl][gh-jrfnl], [@schlessera][gh-schlessera], [@soulseekah][gh-soulseekah], [@twdnhfr][gh-twdnhfr], [#301][gh-301], [#309][gh-309], [#379][gh-379], [#444][gh-444], [#492][gh-492], [#610][gh-610])

- **Removed functionality**

>   The following method, which were deprecated during the 1.x cycle, have now been removed:

Author: ```suggestion
  The following methods, which were deprecated during the 1.x cycle, have now been removed:
```

  - In a number of places, code modernizations, possible now the minimum PHP version has gone up to PHP 5.6, have been applied.
    ([#504][gh-504], [#506][gh-506], [#512][gh-512], [#539][gh-539], [#541][gh-541], [#599][gh-599], [#623][gh-623])

  - Lots of improvements were made to render the tests more reliable and increase the coverage.
    ([#446][gh-446], [#459][gh-459], [#472][gh-472], [#503][gh-503], [#508][gh-508], [#511][gh-511], [#520][gh-520], [#521][gh-521], [#548][gh-548], [#549][gh-549], [#550][gh-550], [#551][gh-551], [#552][gh-552], [#553][gh-553], [#554][gh-554], [#555][gh-555], [#556][gh-556], [#557][gh-557], [#558][gh-558], [#566][gh-566], [#581][gh-581], [#591][gh-591], [#595][gh-595])

  - The move for all CI to GitHub Actions has been finalized. Travis is dead, long live Travis and thanks for all the fish.
    ([#447][gh-447], [#575][gh-575], [#579][gh-579])

  - A GitHub Actions workflow has been put in place to allow for automatically updating the website on releases.
    This should allow for more rapid releases from now on.
    ([#466][gh-466], [#544][gh-544], [#545][gh-545], [#563][gh-563], [#569][gh-569], [#583][gh-583], [#626][gh-626])

>   - Dev dependencies have been updated.

Author: ```suggestion
  - Development-only dependencies have been updated.
```

    ([#516][gh-516], [#517][gh-517])

>   - Various other general housekeeping and improvements for contributors

Author: ```suggestion
  - Various other general housekeeping and improvements for contributors.
```
